docs/development/module/off-chain-workers.md

```diff
@@ -0,0 +1,383 @@
+---
+title: Off-Chain Workers
+---
+
+You can learn about the high level concept of off-chain workers in our [Conceptual Guide](conceptual/core/off-chain-workers.md).
+
+## Using Off-Chain Workers in Runtime
+
+To use off-chain workers in your runtime, include these following modules
+
+**runtime/src/my_runtime.rs**
+
+```rust
+//For better debugging (printout) support
+use support::{ debug, dispatch };
+use system::offchain;
+use sp_runtime::transaction_validity::{
+  TransactionValidity, TransactionLongevity, ValidTransaction, InvalidTransaction
+};
+```
+
+Include the following associated type in your runtime trait for sending signed and unsigned transactions from off-chain worker.
+
+**runtime/src/my_runtime.rs**
+
+```rust
+pub trait Trait: timestamp::Trait + system::Trait {
+  /// The overarching event type.
+  type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+  type Call: From<Call<Self>>;
+
+  type SubmitSignedTransaction: offchain::SubmitSignedTransaction<Self, <Self as Trait>::Call>;
+  type SubmitUnsignedTransaction: offchain::SubmitUnsignedTransaction<Self, <Self as Trait>::Call>;
+}
+```
+
+Now inside the `decl_module!` dispatch functions section, define the following function and this serves as the entry point of the off-chain worker after every block import.
+
+**runtime/src/my_runtime.rs**
+
+```rust
+decl_module! {
+  pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+    // ...
+    // ...
+    fn offchain_worker(block: T::BlockNumber) {
+      debug::info!("Hello World.");
+    }
+  }
+}
+```
+
+By default, the offchain worker doesn't have direct access to user keys (even in development environment), but can only access  app-specific subkeys for security reason. You need to define the `KeyTypeId` at the top of your runtime that is used to group your app-specific subkeys as followed.
+
+**runtime/src/my_runtime.rs**
+
+```rust
+// The key type ID can be any 4-character long string
+pub const KEY_TYPE: KeyTypeId = KeyTypeId(*b"abcd");
+// ...
+// ...
+pub mod crypto {
+  pub use super::KEY_TYPE;
+  use sp_runtime::app_crypto::{app_crypto, sr25519};
+  app_crypto!(sr25519, KEY_TYPE);
+}
+```
+
+To support this runtime, implement in your library module
+
+**runtime/src/lib.rs**
+
+```rust
+/// Define the Transaction signer using the Key definition
+type SubmitTransaction = system::offchain::TransactionSubmitter<
+  my_runtime::crypto::Public, Runtime, UncheckedExtrinsic>;
+
+impl runtime::Trait for Runtime {
```

Comment:
The pallet should not be called `runtime`. How about something like `my_pallet` or `offchain_pallet`?

Reply:
Using `offchain_pallet` now.